File: backend/app/models/maintenance.py
import uuid
from typing import Optional
from datetime import datetime
from sqlalchemy import String, ForeignKey, Enum as SAEnum, DateTime
from sqlalchemy.orm import Mapped, mapped_column, relationship
import enum
from app.models.base import Base, TimestampMixin


class MaintenanceCategory(str, enum.Enum):
    ELECTRICAL = "ELECTRICAL"
    PLUMBING = "PLUMBING"
    FURNITURE = "FURNITURE"
    INTERNET = "INTERNET"
    CLEANLINESS = "CLEANLINESS"
    OTHER = "OTHER"


class MaintenanceStatus(str, enum.Enum):
    SUBMITTED = "SUBMITTED"
    ASSIGNED = "ASSIGNED"
    SCHEDULED = "SCHEDULED"
    IN_PROGRESS = "IN_PROGRESS"
    FIXED = "FIXED"


class MaintenanceRequest(Base, TimestampMixin):
    __tablename__ = "maintenance_requests"

    id: Mapped[uuid.UUID] = mapped_column(primary_key=True, default=uuid.uuid4)
    student_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("students.id"), nullable=False)
    hostel_id: Mapped[uuid.UUID] = mapped_column(ForeignKey("hostels.id"), nullable=False)
    category: Mapped[MaintenanceCategory] = mapped_column(SAEnum(MaintenanceCategory), nullable=False)
    status: Mapped[MaintenanceStatus] = mapped_column(
        SAEnum(MaintenanceStatus), default=MaintenanceStatus.SUBMITTED, nullable=False
    )
    description: Mapped[str] = mapped_column(String(2000), nullable=False)
    photo_url: Mapped[Optional[str]] = mapped_column(String(500), nullable=True)
    room_number: Mapped[str] = mapped_column(String(50), nullable=False)
    assigned_to: Mapped[Optional[uuid.UUID]] = mapped_column(ForeignKey("users.id"), nullable=True)
    scheduled_at: Mapped[Optional[datetime]] = mapped_column(DateTime(timezone=True), nullable=True)
    fixed_note: Mapped[Optional[str]] = mapped_column(String(1000), nullable=True)

    student: Mapped["Student"] = relationship("Student", foreign_keys=[student_id])
    hostel: Mapped["Hostel"] = relationship("Hostel", foreign_keys=[hostel_id])
    assignee: Mapped[Optional["User"]] = relationship("User", foreign_keys=[assigned_to])

File: mobile/packages/dormconnect_core/lib/src/models/maintenance.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'maintenance.freezed.dart';
part 'maintenance.g.dart';

enum MaintenanceCategory {
  @JsonValue('ELECTRICAL') electrical,
  @JsonValue('PLUMBING') plumbing,
  @JsonValue('FURNITURE') furniture,
  @JsonValue('INTERNET') internet,
  @JsonValue('CLEANLINESS') cleanliness,
  @JsonValue('OTHER') other,
}

enum MaintenanceStatus {
  @JsonValue('SUBMITTED') submitted,
  @JsonValue('ASSIGNED') assigned,
  @JsonValue('SCHEDULED') scheduled,
  @JsonValue('IN_PROGRESS') inProgress,
  @JsonValue('FIXED') fixed,
  @JsonValue('CANNOT_FIX') cannotFix,
}

@freezed
class MaintenanceRequest with _$MaintenanceRequest {
  const factory MaintenanceRequest({
    required String id,
    String? hostelId,
    String? reportedById,
    String? reportedByName,
    String? roomNumber,
    required MaintenanceCategory category,
    required MaintenanceStatus status,
    String? title,
    required String description,
    String? location,
    String? assignedToName,
    DateTime? scheduledAt,
    String? fixedNote,
    DateTime? completedAt,
    DateTime? createdAt,
    DateTime? updatedAt,
  }) = _MaintenanceRequest;

  factory MaintenanceRequest.fromJson(Map<String, dynamic> json) =>
      _$MaintenanceRequestFromJson(json);
}
